src/storage/db_interface_frontend.py

```diff
@@ -1,338 +1,427 @@
-            ]}
-            for entry in self.file_objects.find(query, {'_id': 1}):
-                missing_analyses.setdefault(firmware_uid, set()).add(entry['_id'])
+        with self.get_read_only_session() as session:
+            fw_query = self._query_all_plugins_of_object(FileObjectEntry.is_firmware.is_(True))
+            for fw_uid, fw_plugin_list in session.execute(fw_query):
+                fo_query = self._query_all_plugins_of_object(FileObjectEntry.root_firmware.any(uid=fw_uid))
+                for fo_uid, fo_plugin_list in session.execute(fo_query):
+                    missing_plugins = set(fw_plugin_list) - set(fo_plugin_list)
+                    if missing_plugins:
+                        missing_analyses.setdefault(fw_uid, set()).add(fo_uid)
         return missing_analyses
 
+    @staticmethod
+    def _query_all_plugins_of_object(query_filter):
+        return (
+            # array_agg() aggregates different values of field into array
+            select(AnalysisEntry.uid, func.array_agg(AnalysisEntry.plugin))
+            .join(FileObjectEntry, AnalysisEntry.uid == FileObjectEntry.uid)
+            .filter(query_filter)
+            .group_by(AnalysisEntry.uid)
+        )
+
     def find_failed_analyses(self) -> Dict[str, List[str]]:
-        '''Returns a dictionary of failed analyses per plugin: {<plugin>: <UID list>}.'''
-        query_result = self.file_objects.aggregate([
-            {'$project': {'analysis': {'$objectToArray': '$processed_analysis'}}},
-            {'$unwind': '$analysis'},
-            {'$match': {'analysis.v.failed': {'$exists': 'true'}}},
-            {'$group': {'_id': '$analysis.k', 'UIDs': {'$addToSet': '$_id'}}},
-        ], allowDiskUse=True)
-        return {entry['_id']: entry['UIDs'] for entry in query_result}
-
-    def get_data_for_dependency_graph(self, uid, root_uid):
-        data = list(self.file_objects.find(
-            {'parents': uid, 'parent_firmware_uids': root_uid},
-            {'_id': 1, 'virtual_file_path': 1, 'processed_analysis.elf_analysis': 1, 'processed_analysis.file_type': 1, 'file_name': 1})
+        result = {}
+        with self.get_read_only_session() as session:
+            query = (
+                select(AnalysisEntry.uid, AnalysisEntry.plugin)
+                .filter(AnalysisEntry.result.has_key('failed'))  # noqa: W601
+            )
+            for fo_uid, plugin in session.execute(query):
+                result.setdefault(plugin, set()).add(fo_uid)
+        return result
+
+    # --- search cache ---
+
+    def get_query_from_cache(self, query_id: str) -> Optional[dict]:
+        with self.get_read_only_session() as session:
+            entry = session.get(SearchCacheEntry, query_id)
+            if entry is None:
+                return None
+            # FixMe? for backwards compatibility. replace with NamedTuple/etc.?
+            return {'search_query': entry.data, 'query_title': entry.title}
```

Comment:
Rename fields in schema to something sensible. E.g. result and query

Reply:
renamed